[Docs] Terrain Developer Guide (#2122)
o3de/o3de.org: content/docs/user-guide/visualization/environments/terrain/terrain-developer-guide/system-usage.md

The `Default` / `Bilinear` Sampler setting uses the query resolution to sample the four grid corners that surround the requested position, and then it performs bilinear filtering to create the resulting output value. For example, in the illustration below, the underlying height data has different values at every 0.25 meter interval. The chart below it shows what happens when querying this data at position 0.75 with different Height Query Resolution settings.

```goat
40  o                           o
30  |                           |
20  |      o      o             |
10  |      |      |      o      |
0   +------+------+------+------+
   0.00   0.25   0.50   0.75   1.00
```

| Height Query Resolution | Result | Description |
| - | - | - |
| 1.0 meters | 40 | Querying at 0.75 will fetch the underlying data at 0.00 (40) and 1.00 (40) and interpolate at 75%. |
| 0.5 meters | 30 | Querying at 0.75 will fetch the underlying data at 0.50 (20) and 1.00 (40) and interpolate at 50%. |
| 0.25 meters | 10 | Querying at 0.75 will fetch the underlying data at 0.75 (10) and 1.00 (40) and interpolate at 0%. |

> The `Default` Sampler setting is the one that should be used by most systems when querying. However, there are two other sampling methods that are available for specialized use.

Reviewer: ```suggestion
The `Default` Sampler setting should be used by most systems when querying. However, there are two other sampling methods that are available for specialized use.
```

> The `Clamp` Sampler setting will take the requested query position and 'clamp' it down to the terrain grid before requesting the value. Where the `Bilinear` sampler produces an infinite range of interpolated values for all the positions between grid points, the `Clamp` sampler will always produce the same value for every position between grid points. Because this only fetches one data value, not four, it can run up to 4x faster than the `Default` sampler, but the values returned will only match the rendering and physics representations when queried directly on the grid points. For that reason, this setting should only be used when querying directly on terrain grid points.

Reviewer: Small suggestion to recall that there are "four" positions here because it helps quickly answer "why is it 4x faster?"

```suggestion
The `Clamp` Sampler setting takes the requested query position and "clamps" it down to the terrain grid before requesting the value. Whereas the `Bilinear` sampler produces an infinite range of interpolated values for all four positions between the grid points, the `Clamp` sampler always produces the same value for each position. Because this only fetches one data value, not four, it can run up to 4x faster than the `Default` sampler, but the returned values only match the rendering and physics representations when they are queried directly on the grid points. For that reason, this setting should only be used when querying directly on terrain grid points.
```

Author: This might need better clarification, "four" isn't a correct change here. 

Suppose you have four grid points like this where the heights are 5 at X=1, and 6 at X=2:
```
2 +  *5 *6
  |
1 +  *5 *6
  |    
0 +--+--+
   0  1  2
```

What this section is trying to say is that you can query an infinite number of positions between those grid points and get back an infinite number of results. For example, a query of (1.5, 1) gives you a height of 5.5. a query of (1.5001, 1) gives you a height of 5.5001. And so on. That's what "all the positions" is supposed to mean in this context.

In contrast, when using using 'Clamp', a query of (1.5, 1) gives you a height of 5. A query of (1.5001, 1) gives you a height of 5. A query of (1.999, 1) gives you a height of 5. But a query of (2, 1) gives you a height of 6.

Open to suggestions on how to make this more clear...

Author: I've reworked this suggestion and included (hopefully) better diagrams. What do you think?

> The `Exact` Sampler setting will take the requested query position and pass it directly down to the underlying data source, ignoring the terrain grids entirely. Similar to `Clamp`, this is 4x faster than `Default`, but will only produce matching values when queried directly on the grid points. This setting should also only be used when querying directly on terrain grid points, but it may serve some alternate purpose if there's ever a reason to query the underlying terrain data directly without regards to the terrain grid.

Reviewer: My suggestion reworks your last sentence -- I wanted to remove "but" because I think this is an "and" rather than a "but" situation. I also attempt to streamline the last part, such as removing unnecessary words. 
```suggestion
The `Exact` Sampler setting takes the requested query position and passes it directly down to the underlying data source, ignoring the terrain grids entirely. Similar to `Clamp`, this is 4x faster than `Default`, but only produces matching values when it's queried directly on the grid points. You should only use this setting when you are querying directly on the terrain grid points. Alternatively, `Exact` may be useful if you want to directly query the underlying terrain data, without regards to the terrain grid.
```

Author: Hmm, how about this instead?

```suggestion
The `Exact` Sampler setting takes the requested query position and passes it directly down to the underlying data source, ignoring the terrain grids entirely. Similar to `Clamp`, this is 4x faster than `Default`, but only produces matching values when it's queried directly on the grid points. You should not use it to query positions outside of the grid points, since the data may not align with the rendering or physics representations. However, the feature is available for those who want it.
```

The basic points are that the data from `Exact` is useful when queried directly on the grid points, _probably_ isn't useful anywhere else, but _maybe_ someone will have a neat way to use it and we don't want to prevent them from doing it.

Reviewer: Yes, sounds good!

### The `terrainExists` flag

Because terrain data only exists within arbitrary authored regions, it's possible to query the terrain system for points where the terrain does not exist. The terrain is defined as existing as long as height data exists. When querying for terrain data, this flag should always be checked for each query result to ensure that the data actually represents valid terrain.

### Query optimizations

Here are a few tips for getting the best performance out of the terrain query APIs:

> * The APIs are designed to let you query for different combinations of height, normal, and surface data. The more data you query, the more expensive the query will be, so try to limit queries to fetch the minimum amount of data that you intend to use. When you need multiple types of data for the same position, the APIs that let you fetch multiple types at once will generally be faster than calling the individual APIs separately. For example, `GetSurfacePoint` will be faster than calling `GetHeight`, `GetNormal`, and `GetSurfaceWeights` individually.

Reviewer: ```suggestion
* The APIs are designed to let you query for different combinations of height, normal, and surface data. The more data you query, the more expensive the query is, so try to limit queries to fetch the minimum amount of data that you intend to use. When you need multiple types of data for the same position, the APIs that let you fetch multiple types at once are generally faster than calling the individual APIs separately. For example, `GetSurfacePoint` is faster than calling `GetHeight`, `GetNormal`, and `GetSurfaceWeights` individually.
```

* When querying for multiple input positions, the bulk APIs (`QueryList`, `QueryRegion`) are generally 80-90% faster than calling the individual APIs (`GetHeight`, etc).
* `QueryRegion` / `QueryRegionAsync` are the fastest query APIs because they don't require the additional step of building up an input list of positions before calling them.

> * The queries can be made up to 4x faster by using the `Exact` or `Clamp` sampler choices, as long as you take care to align the input positions or input region exactly to the terrain grid. Misalignment will produce inaccurate results.

Reviewer: ```suggestion
* The queries can be up to 4x faster by using the `Exact` or `Clamp` sampler choices, as long as you take care to align the input positions or input region exactly to the terrain grid. Misalignment produces inaccurate results.
```

> * The `TerrainAreaExistsInBounds` provides a quick way to determine if terrain exists at _all_ within an AABB. This can be used to circumvent the more expensive per-point queries entirely if the region your system cares about doesn't contain any terrain.

Reviewer: ```suggestion
* The `TerrainAreaExistsInBounds` provides a quick way to determine if terrain exists within an AABB. This can be used to entirely circumvent the more expensive per-point queries if the region that your system cares about doesn't contain any terrain.
```

### Raycast queries

The `TerrainDataRequestBus` provides a `GetClosestIntersection` method to perform a single raycast against the terrain. Typically, collision raycasts are expected to be performed through a physics system, but this method is available as an alternative for when physics raycasts aren't appropriate. For example, it's used within the Editor itself for placing objects on the terrain.

> If your system is always raycasting straight down to get terrain information directly beneath a location, the query APIs will produce faster results than the raycast API.

Reviewer: I'm assuming the query APIs refers to the ones mentioned earlier on this page, and the raycast API refers to this `GetClosestIntersection` function? If so, I suggest adding that clarity here. I think it starts to get confusing which API you're talking about because this section is called "raycast queries"

```suggestion
If your system often raycasts straight down to get terrain information directly beneath a location, the query APIs that are mentioned earlier produces faster results than this raycast API.
```

Author: Hm, after re-reading this block, how about this instead:

```suggestion
{{< tip >}}
`GetClosestIntersection` provides generalized raycasts in any direction, but if your system only raycasts straight down to get terrain information directly beneath a location, use one of the other query APIs such as `GetSurfacePoint` or `QueryList` instead. They produce faster results.
{{< /tip >}}
```

Reviewer: Yes that sounds clearer! Small note: "generalized" -> "general"

Author: Discussed offline, we'll keep "generalized" since it has an intentional use in this sentence.
